File: aesp/src/main/java/ut/aesp/model/LearnerPackage.java
package ut.aesp.model;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import ut.aesp.enums.PaymentStatus;

import java.math.BigDecimal;
import java.time.LocalDateTime;

import org.hibernate.envers.Audited;

@Getter
@Setter
@Audited
@Entity
@Table(name = "learner_packages")
public class LearnerPackage extends Auditable {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  @ManyToOne
  @JoinColumn(name = "learner_id", nullable = false)
  private LearnerProfile learner;

  @ManyToOne
  @JoinColumn(name = "package_id", nullable = false)
  private Package packageEntity;

  private Integer transactionId;
  private LocalDateTime purchaseDate;
  private BigDecimal priceAtPurchase;
  private LocalDateTime expireDate;

  @Enumerated(EnumType.STRING)
  @Column(length = 20)
  private PaymentStatus paymentStatus = PaymentStatus.PENDING; // pending | completed | failed
}

File: ASEP-FE/src/schemas/LearnerPackage.ts
import { z } from "zod";

export const LearnerPackageSchema = z.object({
  id: z.number().int().optional(), 

  learner_id: z
    .number()
    .int()
    .refine((val) => !isNaN(val), {
      message: "learner_id phải là số hợp lệ",
    }),

  package_id: z
    .number()
    .int()
    .refine((val) => !isNaN(val), {
      message: "package_id phải là số hợp lệ",
    }),

  transaction_id: z.number().int().nullable().optional(),

  purchase_date: z
    .string()
    .refine((val) => !isNaN(Date.parse(val)), {
      message: "purchase_date phải là ISO datetime hợp lệ",
    })
    .optional(),

  price_at_purchase: z
    .number()
    .nonnegative({ message: "Giá phải >= 0" })
    .nullable()
    .optional(),

  expire_date: z
    .string()
    .refine((val) => !isNaN(Date.parse(val)), {
      message: "expire_date phải là ISO datetime hợp lệ",
    })
    .nullable()
    .optional(),

  payment_status: z
    .enum(["PENDING", "COMPLETED", "FAILED"], {
      message: "payment_status phải là một trong: PENDING | COMPLETED | FAILED",
    })
    .default("PENDING"),

  created_at: z
    .string()
    .refine((val) => !isNaN(Date.parse(val)), {
      message: "created_at phải là ISO datetime hợp lệ",
    })
    .optional(),

  updated_at: z
    .string()
    .refine((val) => !isNaN(Date.parse(val)), {
      message: "updated_at phải là ISO datetime hợp lệ",
    })
    .optional(),
});


export type LearnerPackage = z.infer<typeof LearnerPackageSchema>;
